src/btree/bt_sync.c

```diff
@@ -223,7 +265,7 @@ __sync_ref_int_obsolete_cleanup(WT_SESSION_IMPL *session, WT_REF *parent)
         ref = pindex->index[slot];
 
         if (__sync_ref_is_obsolete(session, ref))
-            WT_RET(__sync_ref_evict_or_mark_deleted(session, ref));
+            WT_RET(__sync_ref_evict_or_mark_deleted(session, ref, flags, rlp));
```

Comment:
I know we must have talked about this, but why can't these pages be evicted while we're traversing the parent page's index? We're looking at things like `WT_REF.page` in `__sync_ref_is_obsolete()` without any locking, and I can't remember why that's safe. Could you please add a comment?

Reply:
Actually it is not safe to process the in-memory child page while holding the hazard pointer on the parent page. If the child page is in-memory, it should be already processed during the tree walk, so there is no need to process them at the internal page. So I added a comment to explain it.

Reply:
Here's the path that concerns me:

1. if (ref->state == WT_REF_DISK && __sync_ref_is_obsolete(session, ref))
2. if (ref->addr != NULL &&
3. page is read into memory modified & reconciled, changing WT_REF.addr
4. __wt_ref_is_leaf() is called and is using WT_REF.addr without any locking

Can that not happen?

Reply:
Yes, the above scenario can happen. I rearranged the code to lock the ref and mark it delete before we release the lock, so the ref cannot be changed.